c10/core/DispatchKey.h

```diff
@@ -19,8 +19,53 @@ namespace c10 {
 // we "count leading zeros" when we extract the highest priority dispatch
 // key.)
 //
+//
+
 // NOTE: Keep the list in sync with `DispatchKey` in tools/codegen/model.py
-enum class DispatchKey : uint8_t {
+enum class BackendBit : uint8_t {
+
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~ BACKENDS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+  // A "backend" is colloquially used to refer to handlers for dispatch
+  // which actually implement the numerics of an operation in question.
+  //
+  // Due to the nature of the enum, these backends are specified in
+  // an ordered way, but for most backends this order is not semantically
+  // meaningful (e.g., it's valid to reorder these backends without changing
+  // semantics).  The only situation when backend ordering is meaningful
+  // is when the backend participates in multiple dispatch with another
+  // backend; e.g., CPU and CUDA (cuda must have higher priority).
+
+  // These keys don't correspond to individual kernels.
+  // Instead, they represent the backends that are allowed to override specific pieces of functionality:
+  // - dense kernels (e.g. DispatchKey::CPU)
+  // - sparse kernels (e.g. DispatchKey::SparseCPU)
+  // - quantized kernels (e.g. DispatchKey::QuantizedCPU)
+  // - autograd kernels (e.g. DispatchKey::AutogradCPU)
+  // We reserve space in the runtime operator table for this full cross product of [backends] x [functionality]
+  // [For backends in this list] x [keys below that are explicitly marked as having per-backend functionality]
+
+  InvalidBit = 0,
+  CPUBit,
+  CUDABit,
+  HIPBit,
+  XLABit,
+  MLCBit,
+  XPUBit,
+  HPUBit,
+  VEBit,
+  LazyBit,
+  PrivateUse1Bit,
+  PrivateUse2Bit,
+  PrivateUse3Bit,
+  // Define an alias key to represent end of backend dispatch keys.
+  // If you add new backend keys after PrivateUse3, please also update it here.
+  // (But you shouldn't: private use keys should have higher precedence than
```

Comment:
Yep, although that was true before, for the most part. But you're right that this causes some re-orderings (XLA used to be after ORT, and now it's before). In practice I don't think that this is much of a concern, since the only mixed device code I'm aware of is [non_cpu_backend] x [cpu_scalar].

Reply:
Agreed that this is an ok change.
We should still mention that the "bottom of functionality" is in the middle of DispatchKey I think.

Reply:
(I revamped the ~~~~Functionality Keys~~~~ comment section, and I mention that here)